### PR TITLE
Attribute for generating quickcheck tests with custom derive

### PR DESCRIPTION
**Pull request by WaDelma**
Addresses #5 
This not yet ready for merge: Looking for comments.

So I implemented attribute for generating quickheck tests when deriving Alga traits.
This work pretty well, but there is problem with quasigroup tests: Most (all?) quasigroups don't contain additive identity (that is zero), because it messes with division (dividing by zero).
This is fine in itself as we can just implement it for types that don't contain zero.

Things get more interesting when we get to fields as it requires to contain additive identity to satisfy Monoidity of multiplication.
Afaik it's common to just say that the multiplication needs to be quasigroup for all but zero, but that's bit tricky to express when generating the quickcheck tests.

My first idea was to create wrapper type called `NonIdentity` and implement `Arbitary` for it that doesn't generate identity element and the hard code it to generate it for fields multiplication. But then I become unsure if I actually understand all the nuance of what is happening here.

So does any of this make any sense? Is there any cases I am missing? What should be done here?

### CODE REVIEW

**Comment by sebcrozet** [comment]
I am not sure I understand your problem exactly, so perhaps a concrete example of quickcheck test you generate for some specific type (say, Vec2<f64>) and trait would help.

Here is what I understand. You want to generate a quickcheck test that tests if a property is satisfied for a pair of some type (say, f64). Your test naturally looks like that:

```rust
fn test_property(a: f64, b: f64) -> bool {
  // check the property
}
```

The problem here is that your property says that _« The relation a R b must be satisfied for all  a and all b except its additive identity »_. So the solution you are proposing is to wrap f64 into a type that implements `Arbitrary` which does not generate 0. So you end up with:

```rust
struct NonIdentity<T> {
   // fields
}

impl<T> Arbitrary for NonIdentity<T> {
  // methods implementations.
}

fn test_property(a: f64, b: NonIdentity<f64>) -> bool {
  // check the property.
}
```

I think this is fine but adds a lot of boilerplate code. In particular, other tests might want to ignore other special values (like 1) so you might end up defining even more wrappers. I think a much simpler way of doing that is to read the property as: _« The relation a R b must be satisfied when b != 0. If b == 0, then it is unconditionally satisfied. »_ (Here 0 is to be interpreted as the additive identity.) The corresponding quickcheck test would be:

```rust
fn test_property(a: f64, b: f64) -> bool {
  if b != 0 {
    // check the property.
  }
  else {
    true // Satisfied for any value of a.
  }
}
```

In other words, we could just test the special values inside of the test. Mathematically speaking, we just extended the definition of our property to include the special value 0. Quickcheck will end up trying other nonzero values of `b` anyway!

**Comment by WaDelma** [comment]
For specific example: Currently quickcheck test `prop_inv_is_latin_square_approx_for_Rational_as_AbstractQuasigroup_Multiplicative` for `Rational` fails because of identity element.

**Comment by WaDelma** [comment]
`RationalShrinker` was overflowing before I fixed unrelated problem, so I am pretty sure it's still broken.

**Comment by WaDelma** [comment]
I just realised that this doesn't work with generic types because to generate values of them all generic type parameters need to specified and there is currently no way to do that.§

**Comment by WaDelma** [comment]
Generic types are now supported, but there are some error messages missing and the code is just strewn together.

**Comment by WaDelma** [comment]
I just realised that I had copied the vectors example to tests folder so that it would execute test generated by the attribute.

Should I move the example to tests folder or should I make dummy test file that just links to the example?

**Comment by sebcrozet** [comment]
I think it is better to keep the example in the example folder. That way, people looking for examples will know where to find them. The real issue here comes from cargo not compiling tests on the examples folder so it seems more adequate to just make a workaround (with the dummy test file) and wait for a fix.

**Comment by WaDelma** [comment]
I think that this PR is pretty much ready for review now.

**Comment by sebcrozet** [comment]
This looks pretty good. Though would you like to add at least some minimal documentation regarding the various attributes you added (and their syntax)? I'm not sure if rustdoc has a handy way of attaching documentation to attributes. Otherwise this could go, e.g., on the `lib.rs` of alga (even though currently most of the doc is on `alga/general/mod.rs`).

**Comment by WaDelma** [comment]
I added the documentation to `alga_derive` crate and added mentions that the crate exists to the main `alga` documentation.

**Comment by sebcrozet** [comment]
Thanks!
